### DISCUSSION TITLE
Easy way to create multilingual notebooks

### DISCUSSION THREAD

**Discussion by fonsp** [fonsp/Pluto.jl#1444] [Pluto development]
Since Pluto is all about making scientific computing accessible to more people, we should make it easy for people to create a notebook that contains multiple translations of the contained text.

**Comment by fonsp** [comment]
# Idea 1:

```julia
PlutoUI.multilingual(
	"en" => md"# Hello world!",
	"fr" => md"# Salut le monde !",
)
```

wraps the elements in `span`s and uses some JS to switch the preferred one to visible, based on `navigator.languages` (the user's system language preferences, for me this is `['en-US', 'en', 'nl', 'fr', 'de']`). These are [BCP47 language codes](https://github.com/libyal/libfwnt/blob/b4a75eaacf73834fa964dc29668c70f06caa4913/libfwnt/libfwnt_locale_identifier.c#L28).

**Reply by fonsp** [reply]
## Disadvantage
A big reason to solve this problem is to translate the Getting Started and Basic Mathematics notebook (https://github.com/fonsp/PlutoEducation.jl/issues/1), but this new syntax will make them more complicated. People who want to learn how to write a basic notebook can no longer look at the code of the sample notebooks to learn how.

**Reply by greimel** [reply]
One can overcome this disadvantage by mixing it with idea 2: Have a github action generate single language notebooks (as in idea 2) to be hosted online for beginners to look at. (I guess multi-language support is an advanced issue that can be omitted from intro notebooks)

**Reply by fonsp** [reply]
I have started implementing this in https://github.com/fonsp/disorganised-mess/blob/main/MultiLingual.jl

@samusz and others: Can you try it out and leave feedback? You can use it in another notebook by simply copying the cell that defines `MultiLingual`.

**Comment by fonsp** [comment]
# Idea 2:

We introduce a standardized naming system for notebook files, e.g. `my notebook.jl` with translations `my notebook-lang_fr.jl`, `my notebook-lang_en-CA.jl`. When a notebook is opened (using `/open` or `/sample`), we look at the `Accept-Language` HTTP header to select the translated notebook instead.

**Reply by fonsp** [reply]
## Disadvantage

This will make multi-lingual notebooks more difficult to maintain: once a translation has been created, you will always have to modify both notebooks in parallel.

**Comment by fonsp** [comment]
# Idea 3:

We build this functionality directly into Pluto, i.e. we add a button to write a translated version of a cell, and a way to switch between versions.

**Reply by fonsp** [reply]
## Disadvantage

Requires many new features and a change to our file format.

**Reply by lungben** [reply]
Or add a button that translates automatically all text cells (Markdown, html) to the preferred language using a web translation service (like Google Translate - but I do not know if its API is free to use).

This way, the author would not have to write multiple languages herself.

**Reply by schneiderfelipe** [reply]
Maybe silently wrap all markdown in a notebook in an object that supports language selection (with a fallback when the file is run as a script)? This would avoid changing the file format, but translations in code (anything other than markdown cells) would not be possible.

**Reply by samusz** [reply]
I feel restricting this issue to Pluto may be the wrong way to go, long term wise. 
The i18n/l8n issue creeps everywhere in Julia and l feel it should  be addressed at a lower more Julia core level. This is exemplified by the methods docs, a huge part of the help system.
If we're only addressing the notebooks then a different  lang folder should be enough for starters

**Reply by fonsp** [reply]
@samusz That's a great point! @logankilpatrick and @miguelraz might be able to say more about localisation of julia docstrings. Maybe we should open a new discussion somewhere else about this?

**Reply by fonsp** [reply]
This came up again recently! @logankilpatrick or @miguelraz, has there been any previous work on this?
